pubtools/_quay/utils/stepper.py

```diff
@@ -0,0 +1,354 @@
+
+    To provide detailed info about step status, step can store these details in _details
+    attribute of the step instance. Details can contain anything json compatible. To
+    set details to initial state, user needs to overwrite _init_details.
+    Later, update_details method can be used to updating the details.
+    User can overwrite `_pre_run` method to do any kind of 'lazy' preparation of data or
+    set `skip` and `skip_reason` variables in the instance to prevent step from the
+    execution.
+    When there's data error or wrong data are provided, _run method can
+    raise StepFailedError to set step to 'failed' state. Failed state indicates
+    there's problem with data or configuration and step in this state cannot be
+    executed again. If any other exception occurs and is not caught in _run, step is set
+    to 'error' state and can be executed again. Only other two states indicating step is
+    able to be executed are 'ready' and 'prep'. Ready state is set after initialization
+    of the step instance. Step is switched to 'prep' state just before `_pre_run`
+    is called. If step is in 'prep' state  `_pre_run` is not called again. After step
+    finished the execution, and if there wasn't any error, `results` instance attribute
+    is stored to `shared_results`.
+    Last thing to do is to set NAME class attribute identify type of step
+    """
+
+    NAME = "AbstractSequenceStep"
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(
+        self, uid, step_args, step_kwargs, shared_results, external_resources=None
+    ):
+        """Initilize the step.
+
+        Args:
+            uid: (str)
+                An unique id for indentifying two steps of the same class
+            uid: (step_args)
+                Arguments used for the step
+            uid: (step_kwargs)
+                Dictionary of a name arguments for the step
+            shared_results: (dict-like object)
+                Object to store shared data between steps
+            external_resources: any-object
+                Reference for external resources (in any form) which are constant and
+                shouldn't be part of the step state or step data
+        """
+        self.step_args = step_args
+
+        self.masked_args = [
+            str(arg) if isinstance(arg, Secret) else arg for arg in step_args
+        ]
+        self.masked_kwargs = dict(
+            [
+                (k, str(arg)) if isinstance(arg, Secret) else (k, arg)
+                for k, arg in step_kwargs.items()
+            ]
+        )
+
+        self.step_args = [
+            arg.value if isinstance(arg, Secret) else arg for arg in step_args
+        ]
+        self.step_kwargs = dict(
+            [
+                (k, arg.value) if isinstance(arg, Secret) else (k, arg)
+                for k, arg in step_kwargs.items()
+            ]
+        )
+
+        self.step_kwargs = step_kwargs
```

Comment:
you are overwriting the previous step_kwargs definition

Reply:
ah, right, I forgot to remove it